File: typescript/structual-patterns/proxy.ts
interface IDatabase {
    operation(): void;
}

class Database implements IDatabase {
    operation(): void {
        console.log("Database service operation executed.");
    }
}

class DatabaseProxy implements IDatabase {
    private service: Database

    constructor(service: Database) {
        this.service = service;
    }

    operation(): void {
        if (this.checkAccess()) {
            console.log("Proxy: Access granted. Executing operation.");
            this.service.operation();
        } else {
            console.log("Proxy: Access denied. Operation not executed.");
        }
    }

    checkAccess(): boolean {
        // Simulate access control logic
        console.log("Proxy: Checking access permissions.");
        return true; // Assume access is granted for simplicity
    }
}

function clientCode(service: IDatabase) {
    service.operation();
}

const database = new Database();
const proxy = new DatabaseProxy(database);

clientCode(proxy); // Using the proxy to access the database service

console.log("-----------")

clientCode(database); // Directly using the database service

File: rust/src/bin/structual-patterns/proxy.rs
trait Database {
    fn operation(&self);
}

struct RealDatabase;

impl Database for RealDatabase {
    fn operation(&self) {
        println!("Performing operation on the real database.");
    }
}

struct ProxyDatabase {
    real_database: RealDatabase,
}

impl ProxyDatabase {
    fn new() -> Self {
        ProxyDatabase {
            real_database: RealDatabase,
        }
    }

    fn check_access(&self) -> bool {
        println!("Checking access permissions.");

        true
    }
}

impl Database for ProxyDatabase {
    fn operation(&self) {
        if self.check_access() {
            println!("Access granted. Delegating operation to the real database.");
            self.real_database.operation();
        } else {
            println!("Access denied.");
        }
    }
}

fn call(service: &dyn Database) {
    service.operation();
}

fn main() {
    let db = RealDatabase;
    let proxy = ProxyDatabase::new();

    call(&db);

    println!("-------");

    call(&proxy);
}
